File: internal/app/exec.go
package app

import (
	"bytes"
	"context"
	"errors"
	"fmt"
	"strings"
	"time"

	"github.com/starfield17/rhost/internal/config"
	"github.com/starfield17/rhost/internal/errs"
	"github.com/starfield17/rhost/internal/shell"
	"github.com/starfield17/rhost/internal/transport/openssh"
)

// killTimeout bounds the best-effort remote process-group kill issued after a
// foreground timeout.
const killTimeout = 8 * time.Second

// maxExecOutputBytes keeps one invocation's two in-memory streams bounded even
// when an API caller bypasses the CLI's flag validation.
const maxExecOutputBytes = 64 * 1024 * 1024

// ExecOptions describes a single stateless foreground execution.
type ExecOptions struct {
	Host    string
	Command string
	Cwd     string
	Env     map[string]string
	Timeout time.Duration
	// Stdin is fed to the remote command. It exists for the helpers that take a
	// structured request rather than a path, so no argument ever has to be quoted
	// into a shell (fs read/write/patch use it).
	Stdin []byte
	// MaxOutputBytes bounds each output stream *in the CLI's memory and in the
	// JSON*. 0 means unbounded, which is what `exec` used to do always; the
	// command itself is never truncated, only what rhost carries back.
	MaxOutputBytes int
}

// ExecResult is the outcome of a foreground execution.
//
// The `*_bytes` counters are what the command produced, the `*_truncated` flags
// say whether `Stdout`/`Stderr` are all of it, and `CleanupConfirmed` separates
// "this run is over" from "this run is over and I proved the remote stopped".
type ExecResult struct {
	Host         string
	ExitCode     int
	Stdout       string
	Stderr       string
	Duration     time.Duration
	TimedOut     bool
	Cancelled    bool
	CancelSignal string

	StdoutBytes      int64
	StderrBytes      int64
	StdoutTruncated  bool
	StderrTruncated  bool
	CleanupConfirmed bool
}

// Execute runs a command in a fresh remote execution context.
//
// A non-nil *errs.Error indicates an adapter/transport failure (or a timeout).
// A nil error means the remote command actually ran; ExitCode is then the
// remote command's real status.
func (a *App) Execute(ctx context.Context, opts ExecOptions) (ExecResult, *errs.Error) {
	out := ExecResult{Host: opts.Host, ExitCode: -1}
	if opts.MaxOutputBytes < 0 || opts.MaxOutputBytes > maxExecOutputBytes {
		return out, errs.New(errs.ConfigInvalid,
			"max output bytes must be between 0 and 67108864", false)
	}

	for k := range opts.Env {
		if err := shell.ValidateEnvKey(k); err != nil {
			return out, errs.Wrap(errs.ConfigInvalid, err.Error(), false, err)
		}
	}
	if strings.TrimSpace(opts.Command) == "" {
		return out, errs.New(errs.ConfigInvalid, "no command given", false)
	}

	nonce, err := openssh.NewNonce()
	if err != nil {
		return out, errs.Wrap(errs.Internal, "could not generate nonce", false, err)
	}

	script := openssh.BuildScript(openssh.ExecSpec{
		Command: opts.Command,
		Cwd:     opts.Cwd,
		Env:     opts.Env,
		Nonce:   nonce,
	})

	timeout := opts.Timeout
	if timeout <= 0 {
		timeout = a.DefaultTimeout
	}

	res, runErr := a.SSH.RunWith(ctx, opts.Host, openssh.WrapScript(script), openssh.RunOptions{
		Timeout:        timeout,
		MaxOutputBytes: opts.MaxOutputBytes,
		Stdin:          opts.Stdin,
	})
	if runErr != nil {
		if errors.Is(runErr, config.ErrUnsafeLocalState) {
			return out, errs.Wrap(errs.ConfigInvalid, runErr.Error(), false, runErr)
		}
		return out, errs.Wrap(errs.SSHUnreachable, runErr.Error(), true, runErr)
	}
	out.Duration = res.Duration

	if res.TimedOut {
		out.TimedOut = true
		out.Stdout = string(res.Stdout)
		out.Stderr = string(res.Stderr)
		out.StdoutBytes, out.StderrBytes = res.StdoutBytes, res.StderrBytes
		limitExecOutput(&out, opts.MaxOutputBytes)

		// Killing the local ssh process does NOT terminate the remote command
		// (verified on a real host), so explicitly kill the recorded process
		// group. What that second call cannot prove is the important part: a
		// missing pid file or a failed cleanup is equally consistent with the
		// command still running, so rhost reports the uncertainty instead of
		// claiming the command never started.
		kctx, cancel := context.WithTimeout(context.Background(), killTimeout)
		kres, kerr := a.SSH.Run(kctx, opts.Host, openssh.WrapScript(openssh.KillCommand(nonce)), killTimeout)
		cancel()

		if kerr == nil && !kres.TimedOut && bytes.Contains(kres.Stdout, []byte("killed:")) {
			out.CleanupConfirmed = true
			return out, errs.New(errs.RemoteCommandTimeout,
				fmt.Sprintf("command exceeded timeout %s", timeout), true)
		}
		return out, errs.New(errs.SSHUnreachable,
			"execution deadline exceeded; remote cleanup could not be confirmed, "+
				"so the command may still be running", true)
	}

	body, code, ok := openssh.ParseMarker(res.Stdout, nonce)
	if !ok {
		out.Stdout, out.Stderr = string(res.Stdout), string(res.Stderr)
		out.StdoutBytes, out.StderrBytes = res.StdoutBytes, res.StderrBytes
		limitExecOutput(&out, opts.MaxOutputBytes)
		return out, classifyMissingMarker(res)
	}

	out.ExitCode = code
	out.Stdout = string(body)
	out.Stderr = string(res.Stderr)
	// The counters are of everything the wrapper printed, which includes the begin
	// and completion markers; the body does not. Subtracting what ParseMarker
	// removed keeps stdout_bytes the size of the *command's* output, so a truncated
	// stream is not reported as slightly larger than it was.
	out.StdoutBytes = res.StdoutBytes - int64(len(res.Stdout)-len(body))
	out.StderrBytes = res.StderrBytes
	limitExecOutput(&out, opts.MaxOutputBytes)
	return out, nil
}

// limitExecOutput cuts each stream to the caller's budget, on a rune boundary.
//
// The transport already kept a bounded prefix plus a protocol-sized suffix, so
// the completion marker survives a cap and the exit code is still the command's
// own; what this does is make the *reported* stream obey the limit that was
// asked for, and set the truncation flags from the true byte counts.
func limitExecOutput(out *ExecResult, limit int) {
	if limit <= 0 {
		return
	}
	cut := func(s string) string {
		if len(s) <= limit {
			return s
		}
		b := []byte(s[:limit])
		return string(b[:len(b)-shell.IncompleteUTF8Suffix(b)])
	}
	out.Stdout, out.Stderr = cut(out.Stdout), cut(out.Stderr)
	out.StdoutTruncated = out.StdoutBytes > int64(limit)
	out.StderrTruncated = out.StderrBytes > int64(limit)
}

File: src/app/exec.rs
//! Foreground execution: one shell program, one submission, honest evidence.
//!
//! This is the use-case layer. It decides what a run *means* and hands the CLI a
//! domain [`ExecOutcome`]; it starts no threads of its own and owns nothing
//! after it returns.
//!
//! Two shapes exist because two callers exist: [`execute_stream`] is the
//! command-line path, where bytes must be observable while the remote program
//! runs, and [`execute_captured`] is the helper path (doctor today; files and
//! sessions later), where the whole result is read once the child is gone.
//!
//! The pieces that are not the two entry points are their own files: `outcome`
//! decides what a finished run means and spends the invocation token, `capture`
//! turns bytes into a bounded [`crate::domain::ProcessOutput`], and `cleanup`
//! asks the remote host to stop a process group — reporting only what the
//! attempt proved.

mod capture;
mod cleanup;
mod outcome;

use crate::domain::{CancelSignal, ExecOutcome, InvocationToken, PreExecFailure, ProcessOutput};
use crate::shell;
use crate::transport::openssh::{Client, Streams};
use crate::transport::process::{Keep, StdinSource, Tap};
use crate::transport::protocol::{self, CompletionStream, ExecSpec, Separator};
use std::io::Write;
use std::time::Duration;

use capture::{process_output, truncate};
use outcome::{decide, realize};

/// Upper bound on one invocation's in-memory streams, even when a caller asks
/// for more: the cap bounds what rhost carries back, never what the remote
/// command may produce.
pub const MAX_CAPTURE_BYTES: usize = 64 * 1024 * 1024;

/// Bounds the best-effort remote process-group stop issued after an interruption.
pub const KILL_TIMEOUT: Duration = Duration::from_secs(8);

/// Default per-stream capture for the JSON path. An agent that asks for output
/// gets a bounded amount by default and can name a different one.
pub const DEFAULT_JSON_CAPTURE: usize = 1024 * 1024;

/// The helper path's own bound. A probe that never finishes is a failure, not an
/// open-ended wait.
pub const DEFAULT_HELPER_TIMEOUT: Duration = Duration::from_secs(60);

/// Diagnostics kept even when nothing else is captured: a transport failure has
/// to be classified, and `ssh` says very little.
const DIAGNOSTIC_CAPTURE: usize = 64 * 1024;

pub struct Request<'a> {
    pub host: &'a str,
    pub command: &'a str,
    pub cwd: Option<&'a str>,
    pub env: &'a [(String, String)],
    /// `None` means no deadline: an ordinary foreground run waits until the
    /// remote program finishes (EXEC-002).
    pub timeout: Option<Duration>,
    /// `None` means keep nothing in memory; `Some(0)` means keep everything.
    pub capture: Option<usize>,
    pub fresh: bool,
}

/// How a running child learns that the local invocation should stop, and which
/// signal answered. Two closures rather than one because the poll has to be
/// cheap and the reason has to be exact: `Run::cancelled` says *that* the run
/// was stopped, the signal says whether the process status is 130 or 143
/// (WIRE-004).
#[derive(Clone, Copy)]
pub struct Cancel<'a> {
    poll: &'a dyn Fn() -> bool,
    reason: &'a dyn Fn() -> Option<CancelSignal>,
}

fn never_requested() -> bool {
    false
}

fn never_cancelled() -> Option<CancelSignal> {
    None
}

impl<'a> Cancel<'a> {
    pub fn new(poll: &'a dyn Fn() -> bool, reason: &'a dyn Fn() -> Option<CancelSignal>) -> Self {
        Self { poll, reason }
    }

    fn signal(&self) -> Option<CancelSignal> {
        (self.reason)()
    }

    /// A run nobody may interrupt locally: the helper path, where the caller has
    /// no terminal to lose.
    pub fn none() -> Self {
        Self {
            poll: &never_requested,
            reason: &never_cancelled,
        }
    }
}

#[derive(Debug)]
pub enum ExecError {
    /// The run happened but cannot be described. Not a connectivity claim.
    Internal(String),
}

impl std::fmt::Display for ExecError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Internal(reason) => write!(f, "{reason}"),
        }
    }
}

/// Application-level validation, which runs *after* the CLI has rejected grammar
/// problems. The distinction is load-bearing: an empty `--command` is a usage
/// error, while whitespace-only content reaches here and is `CONFIG_INVALID`
/// (WIRE-006).
fn invalid(request: &Request<'_>) -> bool {
    request.command.trim().is_empty()
        || request
            .capture
            .is_some_and(|limit| limit > MAX_CAPTURE_BYTES)
        || request
            .env
            .iter()
            .any(|(key, _)| !shell::is_valid_env_key(key))
}

fn spec<'a>(
    command: &'a str,
    request: &'a Request<'_>,
    nonce: &'a InvocationToken,
) -> ExecSpec<'a> {
    ExecSpec {
        command,
        cwd: request.cwd,
        env: request.env,
        nonce,
        remote_state_dir: None,
    }
}

/// Runs a foreground command, forwarding its output as it is produced.
///
/// `cancellation` is polled while the child runs, and answers with the signal
/// that ended the local invocation: the reason distinguishes 130 from 143 and
/// must never be guessed after the fact.
pub fn execute_stream(
    client: &Client,
    request: &Request<'_>,
    forward_stdout: Option<Box<dyn Write>>,
    forward_stderr: Option<Box<dyn Write>>,
    stdin: StdinSource<'_>,
    cancellation: Cancel<'_>,
) -> Result<ExecOutcome, ExecError> {
    if invalid(request) {
        return Ok(ExecOutcome::not_started(
            PreExecFailure::Config,
            ProcessOutput::empty(),
            0,
        ));
    }
    let nonce = protocol::new_nonce().map_err(|error| ExecError::Internal(error.to_string()))?;
    let script = protocol::wrap_script(&protocol::build_script(
        &spec(request.command, request, &nonce),
        Separator::Nul,
    ));
    let mut stdout = CompletionStream::new(Separator::Nul, &nonce, forward_stdout, request.capture);
    let mut stderr = Tap::new(
        forward_stderr,
        request.capture.unwrap_or(DIAGNOSTIC_CAPTURE),
        Keep::Prefix,
    );
    let run = client
        .run_streams(
            &Streams {
                target: request.host,
                remote_command: &script,
                timeout: request.timeout,
                stdin,
                fresh: request.fresh,
                cancelled: cancellation.poll,
            },
            &mut stdout,
            &mut stderr,
        )
        .map_err(|error| ExecError::Internal(error.to_string()))?;
    let output = process_output(stdout.body(), stdout.total(), stderr.body(), stderr.total())?;
    let ended = decide(
        client,
        request,
        &nonce,
        &run,
        stdout.completion(),
        &output,
        cancellation,
    );
    realize(nonce, ended, output, run.duration)
}

/// Runs a command and reads both streams once it is over. Used by probes and
/// helpers that need the whole result, never by the `exec` CLI path.
pub fn execute_captured(
    client: &Client,
    command: &str,
    request: &Request<'_>,
    stdin: Option<&[u8]>,
) -> Result<ExecOutcome, ExecError> {
    if invalid(request) {
        return Ok(ExecOutcome::not_started(
            PreExecFailure::Config,
            ProcessOutput::empty(),
            0,
        ));
    }
    let nonce = protocol::new_nonce().map_err(|error| ExecError::Internal(error.to_string()))?;
    let script = protocol::wrap_script(&protocol::build_script(
        &spec(command, request, &nonce),
        Separator::Newline,
    ));
    let limit = request.capture.unwrap_or(0);
    let captured = client
        .run_captured(
            request.host,
            &script,
            request.timeout.unwrap_or(DEFAULT_HELPER_TIMEOUT),
            limit,
            stdin,
            request.fresh,
        )
        .map_err(|error| ExecError::Internal(error.to_string()))?;
    let (stdout_body, code) = match protocol::parse_marker(&captured.stdout, &nonce) {
        Some((body, code)) => (body.to_vec(), Some(code)),
        None => (captured.stdout.clone(), None),
    };
    // The counters describe the wrapper's own stdout, which carries the begin and
    // completion markers; the body does not. Subtracting what the parse removed
    // keeps stdout_bytes the size of the *command's* output, so a truncation flag
    // means "the command produced more than you were given".
    let removed = captured.stdout.len().saturating_sub(stdout_body.len()) as u64;
    let stdout_bytes = captured.stdout_bytes.saturating_sub(removed);
    let output = process_output(
        truncate(stdout_body, limit),
        stdout_bytes,
        truncate(captured.stderr, limit),
        captured.stderr_bytes,
    )?;
    let ended = decide(
        client,
        request,
        &nonce,
        &captured.run,
        code,
        &output,
        Cancel::none(),
    );
    realize(nonce, ended, output, captured.run.duration)
}
